### ISSUE TITLE
Support for pickling trainers

### ISSUE THREAD

**Issue by BenjaminBossan** [huggingface/tokenizers#941]
Right now, it is impossible to pickle or deepcopy trainer objects. E.g.:

```python
# tokenizers.__version__ is '0.11.6'
import pickle
from copy import deepcopy
from tokenizers.trainers import WordPieceTrainer

trainer = WordPieceTrainer(
    vocab_size=30522, special_tokens=["[UNK]", "[CLS]", "[SEP]", "[PAD]", "[MASK]"]
)
pickle.dumps(trainer)  # raises TypeError
deepcopy(trainer)  # raises TypeError
```

I know that pickling has some disadvantages but sometimes it cannot be avoided, e.g. because an existing library uses it.

Since the trainer seems to be implemented directly in Rust, I don't know what the reason for the errors is.

**Comment by Narsil** [comment]
Hi,

What is the library you are using that's forcing you to pickle `trainers` ? It's curiosity at this point.
We can definitely add this.

**Comment by BenjaminBossan** [comment]
Fair question. I'm currently working on sklearn wrappers for tokenizers, so that you can use them as sklearn transformers (this feature might possibly be added to [skorch](https://github.com/skorch-dev/skorch)). The trainer would be part of the transformer and as such, if you want to pickle the transformer, you need to pickle the trainer (sklearn estimators [can all be pickled](https://scikit-learn.org/stable/modules/model_persistence.html), so if it doesn't work, it would be a shame).

More importantly though, sklearn `clone`s estimators in certain situations, e.g. when doing grid search or parallel feature processing. `clone` uses `deepcopy` under the hood, which fails with trainers. I assume the underlying reason is the same as for the pickle failure.

Finally, I can imagine some users wanting to parallelize training using joblib, which also fails.

**Comment by BenjaminBossan** [comment]
The PR I mentioned is now ready: https://github.com/skorch-dev/skorch/pull/841

This is a mitigation I currently use to circumvent the problem:

https://github.com/skorch-dev/skorch/blob/00821fd93e1895917530c997c45381dd480c764d/skorch/hf.py#L316-L323

However, some problems are still not addressed:

https://github.com/skorch-dev/skorch/blob/00821fd93e1895917530c997c45381dd480c764d/skorch/tests/test_hf.py#L245-L249

The issue with `clone` not working is that it is used widely in sklearn, e.g. for `GridSearchCV`, which would consequently fail.

**Comment by Narsil** [comment]
> Fair question. I'm currently working on sklearn wrappers for tokenizers, so that you can use them as sklearn transformers (this feature might possibly be added to [skorch](https://github.com/skorch-dev/skorch)). The trainer would be part of the transformer and as such, if you want to pickle the transformer, you need to pickle the trainer (sklearn estimators [can all be pickled](https://scikit-learn.org/stable/modules/model_persistence.html), so if it doesn't work, it would be a shame).

Ok, trainers should never be part of an item you should share as part of the model IMO.
It's like wanting to embed a whole dataset in the model itself. Only the resulting `tokenizer` is necessary for tokenization to happen (which can already be pickled I think)

> 
> More importantly though, sklearn clones estimators in certain situations, e.g. when doing grid search or parallel feature processing. clone uses deepcopy under the hood, which fails with trainers. I assume the underlying reason is the same as for the pickle failure.
> 
> Finally, I can imagine some users wanting to parallelize training using joblib, which also fails.

That is true and can definitely have its use !
Just FYI, by default tokenizers also parallelize computation, which can become a problem when the user is also using some kind of threading or multiprocessing. using `TOKENIZERS_PARALLELISM=false` can deactivate this so you're back in total control of parallelism within python. Having 2 layers of parallelism will be slow at best (everyone stealing the CPU creating many context switches) and deadlocks at worst (`tokenizers` has tried its best to prevent this, but I cannot guarantee it's deadlock free.. the GIL is a nasty beast, and there's a lot of platform specifics involved)

Just maybe something to be aware of/account for on your users behalf if you can.

**Comment by BenjaminBossan** [comment]
> Ok, trainers should never be part of an item you should share as part of the model IMO.

Yes, the trainer is not really necessary at prediction time. My current workaround is indeed to remove it when the tokenizer is pickled. That way, pickling works and after loading, users can still tokenize (but re-fitting requires a new trainer). Of course, this still doesn't solve the issue of cloning the tokenizer (e.g. sklearn grid search requires this).

My initial idea was not to store the trainer on the sklearn transformer at all. Instead, it could be created ad hoc during fitting and then discarded after training. This would also solve the problem. However, it would require us to predetermine what trainers can be used, taking away customization from the user. Therefore, I prefer the pattern of passing the trainer instance as defined by the user:

```python
hf_tokenizer = Tokenizer(WordPiece(unk_token="[UNK]"))
trainer = WordPieceTrainer(
    vocab_size=30522, special_tokens=["[UNK]", "[CLS]", "[SEP]", "[PAD]", "[MASK]"]
)
sk_transformer = HuggingfaceTokenizer(
    tokenizer=hf_tokenizer,
    trainer=trainer,
)
sk_transformer.fit(data)
sk_transformer.transform(data)
```

This way, the user has much more control over initialization of the trainer or they can pass their own custom trainers. However, it requires us to store the trainer as an attribute on the class, resulting in the pickling problem.

As a compromise, we could allow users to pass `trainer=None`, in which case it is automatically determined during fit and then discarded. If users want to pass their own trainer instance, they would lose the ability to clone the object.

> Only the resulting `tokenizer` is necessary for tokenization to happen (which can already be pickled I think).

Yes, all other pieces (tokenizer, normalizer, etc.) work fine :+1: 

> Just FYI, by default tokenizers also parallelize computation, which can become a problem when the user is also using some kind of threading or multiprocessing.

This is good to know. Probably the best we can do is to document this fact and warn the user about it. We cannot control if they use the tokenizer in a parallelized fashion (say, grid search with `n_jobs>1`). But since the same problem is true for other sklearn estimators (e.g. random forest can parallelize), it should be an issue that most users are aware of.

**Comment by Narsil** [comment]
> My initial idea was not to store the trainer on the sklearn transformer at all. Instead, it could be created ad hoc during fitting and then discarded after training. This would also solve the problem. However, it would require us to predetermine what trainers can be used, taking away customization from the user. Therefore, I prefer the pattern of passing the trainer instance as defined by the user:

I perfectly understand. The main caveat I want to convey is since `trainers` carry a lot of state during training, pickling them won't carry that state over (and we might not even want  to do that, since those state can be very large). The word pairs for instance are not part of the Trainer, but only created when doing a training, so if you start a training job on a process, pickle it, send it somewhere else, essentially your state will be gone.
The Trainer, are essentially only the parameters used to train your "tokenizer".

I think this behavior is OK, but it might be surprising in the Python world, where pickle tends to store all the state no matter what.

> This is good to know. Probably the best we can do is to document this fact and warn the user about it. We cannot control if they use the tokenizer in a parallelized fashion (say, grid search with n_jobs>1). But since the same problem is true for other sklearn estimators (e.g. random forest can parallelize), it should be an issue that most users are aware of.

Perfect, if issues are ever raised about deadlocks, on your side please feel free to issue here, we might be able to at least crash nicely instead of deadlocking (which is really a bad experience)

**Comment by Narsil** [comment]
@BenjaminBossan 

I created a PR to pickle Trainers.

If you want to test right away, you can checkout the branch and install from source.

```bash
# Install rust tool chain : https://www.rust-lang.org/
pip install setuptools_rust
pip install -e .   # `python setup.py install` might work better depending on your environment

# Try it out !
```

**Comment by BenjaminBossan** [comment]
Thanks a lot @Narsil, this makes me understand the inner workings much better.

I have made a change on my side which allows us to avoid having the trainer instance stored on the object in most circumstances. For this, I made use of the `get_trainer` method on the models, which I was not aware of before. This way, users don't need to pass the trainer instance anymore, as they can be retrieved when needed.

Still, for those circumstances where users want/need to pass instantiated trainers, it will be useful to be able to pickle them. Thanks a lot for working on it.

> If you want to test right away, you can checkout the branch and install from source.

I did not have time to test this yet but will take a look soon.

**Comment by BenjaminBossan** [comment]
@Narsil I tried out this branch and indeed it solves the initial issue. Moreover, the unit tests in my PR that are currently `xfail`ing pass with this branch. Thanks for your support.

**Comment by Narsil** [comment]
Glad it can work, this will probably land in the next release.

We don't have a tight schedule but expect a few weeks. don't hesitate to ping if we can align better with your own release for instance.

**Comment by BenjaminBossan** [comment]
> We don't have a tight schedule but expect a few weeks. don't hesitate to ping if we can align better with your own release for instance.

Thanks. As long as the next tokenizers release is earlier than the next skorch release, which is very likely, there should be no problem.
